src/number_theory/legendre_symbol/auxiliary.lean

```diff
@@ -30,18 +34,194 @@ begin
+  { rw [hh, (by norm_cast : ((2 : ℤ) : R) = 2)] at h', exact ring.two_ne_zero hR h', },
+  { rw [hh, (by norm_cast : ((2 : ℤ) : R) = 2)] at h'', exact ring.two_ne_zero hR h'', },
+end
+
 /-- If `ring_char R = 2`, where `R` is a finite reduced commutative ring,
 then every `a : R` is a square. -/
+-- SUGGESTION: move this to `algebra.char_p.basic`, after `frobenius_inj`
+-- ALTERNATIVE: move to `algebra.char_p.two` (and renane to `char_two.is_square`?)
 lemma is_square_of_char_two' {R : Type*} [fintype R] [comm_ring R] [is_reduced R] [char_p R 2]
  (a : R) : is_square a :=
 exists_imp_exists (λ b h, pow_two b ▸ eq.symm h) $
   ((fintype.bijective_iff_injective_and_card _).mpr ⟨frobenius_inj R 2, rfl⟩).surjective a
 
+/-- A prime `p` is a unit in a finite commutative ring `R`
+iff it does not divide the characteristic. -/
+-- The following three lemmas depend on `is_unit` (from `algebra.group.units`)
+-- and `ring_char` (from `algebra.char_p.basic`).
+-- It seems that neither of these files imports the other. So:
+-- ALTERNATIVE 1: move them to `algebra.group.units` and add the import there
+--                con: this file talks about monoids, not rings
+-- ALTERNATIVE 2: move them to `algebra.char_p.basic` and add the import there
+-- ALTERNATIVE 3: leave them here for the time being
+lemma is_unit_iff_not_dvd_char (R : Type*) [comm_ring R] [fintype R] (p : ℕ) [fact p.prime] :
+  is_unit (p : R) ↔ ¬ p ∣ ring_char R :=
+begin
+  have hch := char_p.cast_eq_zero R (ring_char R),
+  split,
+  { intros h₁ h₂,
+    rcases is_unit.exists_left_inv h₁ with ⟨a, ha⟩,
+    rcases h₂ with ⟨q, hq⟩,
+    have h₃ : ¬ ring_char R ∣ q :=
+    begin
+      rintro ⟨r, hr⟩,
+      rw [hr, ← mul_assoc, mul_comm p, mul_assoc] at hq,
+      nth_rewrite 0 ← mul_one (ring_char R) at hq,
+      exact nat.prime.not_dvd_one (fact.out p.prime)
+             ⟨r, mul_left_cancel₀ (char_p.char_ne_zero_of_fintype R (ring_char R)) hq⟩,
+    end,
+    have h₄ := mt (char_p.int_cast_eq_zero_iff R (ring_char R) q).mp,
+    apply_fun (coe : ℕ → R) at hq,
+    apply_fun ((*) a) at hq,
+    rw [nat.cast_mul, hch, mul_zero, ← mul_assoc, ha, one_mul] at hq,
+    norm_cast at h₄,
+    exact h₄ h₃ hq.symm, },
+  { intro h,
+    rcases nat.is_coprime_iff_coprime.mpr ((nat.prime.coprime_iff_not_dvd (fact.out _)).mpr h)
+      with ⟨a, b, hab⟩,
+    apply_fun (coe : ℤ → R) at hab,
+    push_cast at hab,
+    rw [hch, mul_zero, add_zero, mul_comm] at hab,
+    exact is_unit_of_mul_eq_one (p : R) a hab, },
+end
+
+/-- The prime divisors of the characteristic of a finite commutative ring are exactly
+the prime divisors of its cardinality. -/
+lemma prime_dvd_char_iff_dvd_card {R : Type*} [comm_ring R] [fintype R] (p : ℕ) [fact p.prime] :
+  p ∣ ring_char R ↔ p ∣ fintype.card R :=
+begin
+  split,
+  { exact λ h, nat.dvd_trans h (int.coe_nat_dvd.mp ((char_p.int_cast_eq_zero_iff R (ring_char R)
+                (fintype.card R)).mp (char_p.cast_card_eq_zero R))), },
+  { intro h,
```

Comment:
You can use dot notation on `h : a ∣ b`: `h.trans : b ∣ c → a ∣ c`. You can also inline `split` and `intro` calls within a `refine`.
```suggestion
  refine ⟨λ h, h.trans $ int.coe_nat_dvd.mp $ (char_p.int_cast_eq_zero_iff R (ring_char R)
    (fintype.card R)).mp $ char_p.cast_card_eq_zero R, λ h, _⟩,
```

Reply:
OK.